### PR TITLE
docs: Fallback to reference description for responses

### PR DESCRIPTION
**Pull request by Nelspike**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

This PR attempts to resolve the issue on https://github.com/fastify/fastify-swagger/issues/795 by resolving the `$ref` given in the schema and fetching that description in case there's none. PTAL! 🚀

### CODE REVIEW

File: lib/spec/openapi/utils.js [hunk 1]
```diff
@@ -326,8 +338,13 @@ function resolveResponse (fastifyResponseJson, produces, ref) {
       statusCode = statusCode.toUpperCase()
     }
 
+    const referenceJsonSchema = findReference(definitions, rawJsonSchema.$ref)
```
Comment (jean-michelet) [inline lib/spec/openapi/utils.js:341]: I think this search should happen only if `rawJsonSchema.description` is falsy, not systematically.

File: lib/spec/openapi/utils.js [hunk 2]
```diff
@@ -313,6 +324,7 @@ function resolveResponse (fastifyResponseJson, produces, ref) {
   const responsesContainer = {}
 
   const statusCodes = Object.keys(fastifyResponseJson)
+  const definitions = ref.definitions().definitions
```
Comment (jean-michelet) [inline lib/spec/openapi/utils.js:327]: Can be done in the find function.

File: README.md [hunk 3]
```diff
@@ -468,6 +468,9 @@ fastify.get('/responseDescription', {
   }
 }, () => {})
 ```
+
+Additionally, if you provide a `$ref` in your response schema but no description, if your `$ref` base schema includes a description, that will be used.
```
Comment (jean-michelet) [inline README.md:472]: I am not sure that's very clear. Do you think you should rephrase it?

**Review by jean-michelet — COMMENTED** [review]
Thanks for the PR!

I don't know the code base so I can't approve myself.

**Comment by jean-michelet** [comment]
Does it solve your bug @melroy89?

**Comment by melroy89** [comment]
I directly let everything fall from my hands and looked into this PR! And using this utils.js file from the PR and override the file in `node_modules/@fastify/swagger/lib/spec/openapi/utils.js`, so I could directly test it on my existing code-base.

I'm using `addSchema()` method for shared schemas ([see docs](https://fastify.dev/docs/latest/Reference/Validation-and-Serialization/#adding-a-shared-schema)). Together indeed with `@fastify/swagger` with OpenAPI spec v3 and the `@fastify/swagger-ui` package for the swagger UI documentation you see below.

Too bad the issue with fallback description is still present 😢 :

image

Here such an example of a shared schema.

```js
export const notFoundErrorSchema = {
  $id: 'notFoundError',
  description: 'Not Found',
  type: 'object',
  properties: {
    message: { type: 'string' }
  },
  example: {
    message: 'Not found'
  }
}
```

Again, added via `fastify.addSchema(notFoundErrorSchema)` method.

I setup Fastify swagger like this:

```ts
 await fastify.register(fastifySwagger, {
      mode: 'dynamic',
      openapi: {
        openapi: '3.0.0',
        info: {
          title: 'Backend API Docs',
          description: 'RESTful backend API',
          version: '0.3.0',
        },
        servers: [
           ...
       ],
       tags: [
          ...
       ]
    }
})
```

Fastify Swagger UI is very simple to setup as well, basic is:

```ts
fastify.register(fastifySwaggerUi, {
      routePrefix: '/docs'
})
```

TLDR; issue is still present with fallback reference description in responses.

File: lib/spec/openapi/utils.js [hunk 4]
```diff
@@ -303,6 +303,17 @@ function resolveCommonParams (container, parameters, schema, ref, sharedSchemas,
...
+// Keys to references could be different than the name of the $ref,
+// for example `def-0`
+function findReference (definitions, id) {
+  const schemaKey = Object.keys(definitions).find(key => {
+    const schema = definitions[key]
+    return schema.$id === id
+  })
+
+  return definitions[schemaKey] ?? {}
+}
```
[the remark this answers is not in the document]
Comment (Nelspike) [reply lib/spec/openapi/utils.js:315]: Using `.values` makes sense actually, I'll change that 💪🏼 As for renaming the function, we could also go with that. My objective was having a function where you could resolve the reference for any other kinds of needs within this file for the future, but I can see how that's looking too far ahead and how it can be optimised.

Will make that change to return the description directly 👍🏼

File: lib/spec/openapi/utils.js (hunk 1 above)
Comment (Nelspike) [reply lib/spec/openapi/utils.js:341]: Since we don't need anything else, can definitely do so for now.

File: README.md (hunk 3 above)
Comment (Nelspike) [reply README.md:472]: I can try, how's about something along the lines of:

```
Additionally, if you provide a `$ref` in your response schema but no description, the reference's description will be used as a fallback.
```

**Comment by Nelspike** [comment]
First of all, @jean-michelet, thank you for the review! 🚀 

Secondly and equally thankful to @melroy89 for the detailed reproduction description 💪🏼 From what I can see from your code, the only reason why this wouldn't work is due to the trailing `#` on your `$ref`, which is `'notFoundError#'`. That `$ref` is absolutely correct, so we need to find a way to resolve these properly in this PR. If you want to give it a try without the trailing `#`, I think that would do the trick, at least for this example!

**Comment by Nelspike** [comment]
@melroy89 - That's exactly what I meant in my reply, we need to consider those use cases because your reference is 100% correct 😅 The code I pushed in the PR didn't take that into account yet, but just pushed a change that does! I never meant that you should remove `#`, but that the code I had previously worked without it.

**Comment by Nelspike** [comment]
Seems like the second use case you mentioned on your `EDIT` is blocked by this, @melroy89: https://github.com/fastify/fastify-swagger/pull/676

At the moment, seems like `definitions` is not supported within OpenAPI definitions on `fastify-swagger`, which difficults using refs like `commonResponse#/definitions/notFound`. _However_, your first use case with `#` at the end should be covered by this PR now 👍🏼 Once the other one's merged, it should be supported out of the box!

cc @asc11cat @Eomm @ivan-tymoshenko, tagging you folks for visibility since you were discussing on the other PR 😄

**Comment by melroy89** [comment]
Yup works now! Thank you so much! I was looking for this for years.

@Nelspike I would like thank you by providing a donation! Any payment method will do.

File: lib/spec/openapi/utils.js
Comment (melroy89) [inline lib/spec/openapi/utils.js:1]: What about the lib/spec/swagger/utils.js? Can this be added there as well easily? so OpenAPI folder is for OpenAI spec v3 basically. And the swagger folder is also 'OpenAPI' spec, but for older specifications like OpenAPI v2.

Comment (Nelspike) [reply lib/spec/openapi/utils.js:1]: I can quickly take a look and see if we need it! 👍🏼

**Comment by Nelspike** [comment]
@melroy89 - Thank you for your generosity and recognition! I'm totally okay doing this without a donation whatsoever, happy to help wherever I can 🎉

File: lib/spec/openapi/utils.js
Comment (Nelspike) [reply lib/spec/openapi/utils.js:1]: And it's added on the latest commit 🎉

Comment (melroy89) [reply lib/spec/openapi/utils.js:1]: Nice!

**Comment by melroy89** [comment]
> @melroy89 - Thank you for your generosity and recognition! I'm totally okay doing this without a donation whatsoever, happy to help wherever I can 🎉

OK, if you change your mind. Let me know.

**Comment by melroy89** [comment]
Now hopefully this PR gets merged somewhere before 2030 :)

File: lib/util/resolve-schema-reference.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+'use strict'
+
+function resolveSchemaReference (rawSchema, ref) {
+  const resolvedReference = ref.resolve(rawSchema, { externalSchemas: [ref.definitions().definitions] })
+
+  // Ref has format `#/definitions/id`
+  const schemaId = resolvedReference?.$ref?.split('/')[2]
```
Comment (gurgunday) [inline lib/util/resolve-schema-reference.js:7]: Nit:

```suggestion
  const schemaId = resolvedReference?.$ref?.split('/', 3)[2]
```

**Comment by jean-michelet** [comment]
> Now hopefully this PR gets merged somewhere before 2030 :)

I don't think this type of comment is going to help us move any faster. If @climba03003 is ok with the code, I will merge it.

**Comment by melroy89** [comment]
> > Now hopefully this PR gets merged somewhere before 2030 :)
> 
> I don't think this type of comment is going to help us move any faster. If @climba03003 is ok with the code, I will merge it.

No one can take a joke anymore.

**Review by gurgunday — APPROVED** [review]
lgtm

Haha don't worry @melroy89, I assure you we'll get this done this year :)

I just need another member (@climba03003 would be golden) to review

File: README.md [hunk 6]
```diff
@@ -468,6 +468,9 @@ fastify.get('/responseDescription', {
   }
 }, () => {})
 ```
+
+Additionally, if you provide a `$ref` in your response schema but no description, the reference's description will be used as a fallback.
```
Comment (climba03003) [inline README.md:472]: Please state the `$ref` currently only support the format of matching `$id`.
It cannot lookup the fragment part.

**Review by climba03003 — CHANGES_REQUESTED** [review]

File: lib/util/resolve-schema-reference.js (hunk 5 above)
Comment (Nelspike) [reply lib/util/resolve-schema-reference.js]: Oh yeah, nice catch! Thanks for the update 🎉

File: README.md [hunk 7]
```diff
@@ -468,6 +468,9 @@ fastify.get('/responseDescription', {
   }
 }, () => {})
 ```
+
+Additionally, if you provide a `$ref` in your response schema but no description, the reference's description will be used as a fallback. Note that at the moment, `$ref` will only be resolved by matching with `$id` and not through complex paths. Currently being addressed in [this Pull Request](https://github.com/fastify/fastify-swagger/pull/676).
```
Comment (climba03003) [inline README.md:472]:
```suggestion
Additionally, if you provide a `$ref` in your response schema but no description, the reference's description will be used as a fallback. Note that at the moment, `$ref` will only be resolved by matching with `$id` and not through complex paths.
```

I bit later to say, but you don't need to point a PR in document. We don't knows which exact PR will resolve the issue in future. Other parts is good.

**Review by gurgunday — APPROVED** [review]
lgtm

I can release tonight, if anyone's free, feel free to merge and release

**Review by climba03003 — APPROVED** [review]

**Comment by jean-michelet** [comment]
Release tonight plz @gurgunday!
